### ISSUE TITLE
Standalone Monocle Layout

### ISSUE THREAD

**Issue by eadwu** [hyprwm/Hyprland#2050]
Main reason is to view status bars. Fullscreen provides a monocle experience but in the process takes over the space for status bars.

Moving `eww` over to the `overlay` layer keeps it shown everywhere, but it no longer reserves any space even with `:exclusive`.

I do not think any window rule allows this to happen (preserve reserved space by other windows in fullscreen).

**Comment by kcrmson** [comment]
You'll want to try using "hyprctl dispatch fullscreen 1" on a floating window and it should keep the status bar visible while remaining in fullscreen mode. I just verified it now using qutebrowser (once made floating) but it didn't seem to affect it when tiled.

**Comment by eadwu** [comment]
Nice, that gets it closer. Introduces a problem when you actually try to enter fullscreen though, i.e. on a video player.

I rather think that is actually a bug instead of actually working like that.

I'll eventually look at the codebase to give it a stab at implementing myself since this is a layout I do quite need to go about... just a couple few windows of a browser but just sticking it out there in case it is easy to implement (though I'm not sure how intertwined layouts are in wayland given that there are only 2 still).

**Comment by vaxerski** [comment]
no_gaps_when_only + one group?

**Comment by eadwu** [comment]
One group will provide a standalone monocle layout. It is not compatible with `cyclenext` and `swapnext` (with `changegroupactive` being a replacement for `cyclenext` and `swapnext` not having any equivalent from the dispatchers).

If I am interpreting the custom reserved area correctly, that will mean it will still be reserved when I really want fullscreen, i.e. watching a movie.

For my intents `togglegroup` will be enough to mimic my daily workflow from Xorg tiling managers (should be covered by monocle and dwindle), I will not be mixing groups with non-monocle layouts so binding `changegroupactive` to the same action I have for `cyclenext` should work without issue.

I'm still looking into the docs to figure out the best way to move all windows in the workspace into or out of a group but it should be possible whether I have to define a plugin or invoke `hyprctl`.

**Comment by vaxerski** [comment]
> If I am interpreting the custom reserved area correctly, that will mean it will still be reserved when I really want fullscreen, i.e. watching a movie.

No. It works like a bar, but for bars that don't do that by themselves.

>  I will not be mixing groups with non-monocle layouts so binding changegroupactive to the same action I have for cyclenext should work without issue.

You can write a simple script for this

**Comment by misuzu** [comment]
This is the only feature that I'm using on dwm that's missing, everything else is there. `fullscreen, 1` is not the same.

**Comment by charbelnicolas** [comment]
This is the only feature that is keeping me from switching from awesome wm. I'd love to have a native hyprland monocle implementation.

**Comment by KunaPrime** [comment]
absolutely, this is feature that is essential for most of my workflows

**Comment by adwaymalhotra** [comment]
I miss the different margin values for monocle mode from my BSPWM days. I want a larger margin for monocle mode

Also, if we're doing feature requests, I would also like something like the Stage Manager feature from MacOS where the non-visible windows are "minimized" to the side or bottom when we're in this monocle mode with small thumbnails. It is sort of like a master layout with the master centered but the slave windows don't resize (resizing messes up the content rendering within the window).

**Comment by zelfroster** [comment]
@vaxerski sorry for pinging, any thoughts about implementing this?

**Comment by neuromagus** [comment]
>my thoughts? “No, should I write a plugin?”

all tiled WMs I worked with before using DWL (DWM, Xmonad) have Monocle mode (in Haskell's full screen mode). IMHO, this behavior should be added to the main mode and given the opportunity to choose.

I use many monitors every day, one in the center is always in Monocle, the rest switch to the Tile layout.  
At home, I use one large 60-inch monitor with customizable workspaces set to Monocle or Tile mode (if I need to see more than one window). And this workplace is widespread among my friends...


![workspace(1)](https://github.com/hyprwm/Hyprland/assets/63796667/02157e9f-9845-410c-92ef-5744be202679)

**Comment by vaxerski** [comment]
why not make a windowrule to maximize every tiled window? would behave the same I believe

**Comment by neuromagus** [comment]
>why not make a windowrule to maximize every tiled window? would behave the same I believe

Do you have an example of these settings? I'll try to compare the user experience with Hypr and, for example, DWM with Monocle.

**Comment by eadwu** [comment]
Just realized I never posted what I use, I'm not sure if this all the configuration but it looks right by this comment I left there, the following works in 99% of cases - the most glaring issue is that it is still virtually tiled, so switching windows always imposes a resize operation - which is an sometimes an annoying delay to wait for.

Full screen should be able to take over any reserved space.

```
# Reset the layouts on the current workspace to default (monocle-ish)
bind = $mainMod SHIFT, space, exec, hyprctl dispatch fullscreen 1
bind = $mainMod SHIFT, space, exec, togglegroup
```

**Comment by Ferk** [comment]
> the most glaring issue is that it is still virtually tiled, so switching windows always imposes a resize operation - which is an sometimes an annoying delay to wait for.

Yes... the animation when switching "fullscreen 1" windows is annoying, specially because it actually does make some windows change the content/internal layout as they are being resized.

I was expecting to be able to get rid of this form of animation tweaking the [Animations configuration](https://wiki.hyprland.org/Configuring/Animations/), but I did not manage to find which setting would I need to set to disable this. Does anybody know?

This is all a very awkward workaround. A monocle style layout being properly supported would be best.

**Comment by zakk4223** [comment]
I made a monocle layout: https://github.com/zakk4223/hyprlandMonocle

The layout sets all windows to the monitor/workspace size, honoring reserved space.
The currently focused window is set to maximized. This is to work around the fact you cannot have overlapping tiled windows in Hyprland without introducing focus issues.

Setting the window sizes to the maximized size should prevent resizing apps on window switching.

You can navigate between windows using Hyprland's standard cyclenext or movefocus dispatchers.

If you want to run monocle only on one workspace/monitor, see: https://github.com/zakk4223/hyprWorkspaceLayouts

**Comment by vaxerski** [comment]
looks like a fullscreen group with extra steps

**Comment by zakk4223** [comment]
Using a group for this has issues:

1) windows in groups cannot be navigated with the normal dispatchers (cyclenext, etc). You have to switch to the group navigation ones. I guess you could do some submap twiddling, but now you have to track which layout is active, you have to deal with the active workspace being changed etc. 

2) if you want to gather all windows on the current workspace into a group you have to...I guess write a shell script?

Either of these are probably manageable if you're using stock Hyprland, but if you're using extra layouts, per workspace/monitor layouts etc it's not going to be as easy to do.

You can dynamically switch to this monocle layout if you use my workspacelayout plugin and bind a key to send the 'layoutmsg setlayout monocle' dispatch. (I think I have some ideas to make layout switching better in that plugin too...). However you still have the window order issue...

I don't think the 'preserving window order' thing is doable in Hyprland without each individual layout being forced to implement monocle themselves. Layouts can have an internal concept of window order. They _can_ manipulate hyprland's window order to some degree but none of them currently do this.

**Comment by TAforever** [comment]
I also want the monocle layout to be implemented in hyprland, but since configuration options such as `new_window_takes_over_fullscreen` and `exit_window_retains_fullscreen` have been added, the behavior of monocle can be easily emulated with a simple script. The disadvantage of this is that the windows change size when switching
```bash
#!/bin/bash

fullscreenstatus=$(hyprctl activewindow | grep fullscreenClient | awk '{print $2}')
if [ $fullscreenstatus != 1 ]; then
    hyprctl keyword misc:new_window_takes_over_fullscreen 1
    hyprctl keyword misc:exit_window_retains_fullscreen true
    hyprctl dispatch fullscreen 1
else
    hyprctl keyword misc:new_window_takes_over_fullscreen 2
    hyprctl keyword misc:exit_window_retains_fullscreen false
    hyprctl dispatch fullscreen 1
fi
```

**Comment by saipavanc** [comment]
Just adding a comment to bump up this request. Monocle mode is really useful for my workflow as well, and has several use cases when it makes things a lot more efficient. I did see some options where one can use things like pypr to change into monocle mode, but it is not easy to get back to a neatly tiled form. Can anyone mention if there is any technical difficulty in storing the tiled layout and applying it again when switching out of monocle mode? I am not sure I understand what makes this challenging here, as it is implemented in many other tiling managers.

**Comment by vaxerski** [comment]
You can use `fullscreen,1` and `cyclenext` for a more-or-less monocle mode. For more advanced, you can write your own script or plugin.

**Comment by zakk4223** [comment]
The fullscreen thing works ok but it causes resize events when windows swap in and out of focus, which some people would like to avoid.

I've tried to do it with a layout that keeps all the windows the same size, but hyprland reacts poorly to tiled layouts with overlapping windows.  I think that's probably fixable with hooking a few functions but I hate maintaining plugins that hook really deep+important functions because now that's one more thing I have to pay attention to.

Even with a plugin you probably need an additional plugin for per monitor/workspace layouts. But even then there's no sane way to have a workflow where you have a tiled workspace, then switch it to monocle and then switch back to the tiled workspace with all state preserved. Switching layouts 'removes' windows from the existing layout and thus destroys any state.

If you're ok with windows getting resized on focus change, then just do the fullscreen+cyclenext thing. Otherwise there's no real solution at the moment.

**Comment by vaxerski** [comment]
the resizing is a separate issue that some people want "fixed" anyways.

Why not `->setHidden()` for windows not visible? I think that would fix most of the problems.

**Comment by zakk4223** [comment]
hidden windows get skipped by cycleNext etc.

**Comment by vaxerski** [comment]
`myMonoclePlugin:cycleNext`

**Comment by zakk4223** [comment]
Oh hmm, I'll try the narnia thing. 

Eh, custom dispatchers for layout window management are annoying, I'd rather use the global one and have it work sanely everywhere if possible. I actually use multiple layouts so having different binds is bleh.

Could it just be changed so layouts get the ability to override get(Prev|Next)WindowOnWorkspace? Like just move that from the compositor to IHyprLayout? It looks like the only things that use those functions are cycle and swap anyways.

That or make g_pCompositor:m_vWindows exclusively for rendering and have another window list for swap/cycle window stack manipulation. Right now anything that uses changeWindowZOrder completely messes up window cycle order

**Comment by vaxerski** [comment]
Hello there!

This issue has been closed, as we are moving from issues to discussions. More about this can be found [here](https://wiki.hyprland.org/Contributing-and-Debugging/Issue-Guidelines/) and in #9854.

Firstly, please make sure this issue is still relevant in latest Hyprland. If so, we ask you to open a [discussion](https://github.com/hyprwm/Hyprland/discussions) (please read the discussion guidelines first, in the pinned post)

Thank you for your understanding!

**Comment by noctuid** [comment]
Scrolling layout (e.g. hyprscrolling plugin) is basically a superset of monocle functionality
